src/gui/plugins/scene_manager/GzSceneManager.cc

```diff
@@ -92,25 +147,39 @@ bool GzSceneManager::eventFilter(QObject *_obj, QEvent *_event)
   if (_event->type() == gui::events::Render::kType)
   {
     this->dataPtr->OnRender();
+    if (this->dataPtr->sameProcess)
+    {
+      this->dataPtr->eventManager->Emit<ignition::gazebo::events::Render>();
```

Comment:
If we rely on the GUI to drive the sensor's render loop, I think it will be difficult to lockstep sensors and physics.

I'm not sure what's the best way to go about this. I think that the sensors system should be the one rendering the scene at every iteration, and the GUI should run in a separate process and paint that rendered scene to the window at a lower frame rate. I think that moving the paint code to `ign-gui`'s `MinimalScene` may complicate this even further :confused: 

I don't know if you have any good ideas here, @iche033 , you've been doing a lot of work with the rendering thread.

Reply:
theoretically speaking, it may be possible to have one main rendering loop in a thread shared by sensors and gui. 

so before these changes, we have two threads for rendering, one on server side and one on the GUI side:
* server:  This [RenderThread](https://github.com/ignitionrobotics/ign-gazebo/blob/84956cf51f01414b16527aa6c8b0b0a98535feb8/src/systems/sensors/Sensors.cc#L292) in sensors system (separate thread from the main ECS thread)
* gui: This [RenderThread](https://github.com/ignitionrobotics/ign-gui/blob/6d3690db7cb072664a7382752654763c7cee5a34/src/plugins/minimal_scene/MinimalScene.hh#L172) in Scene3D

What if we make them (merge?) the same thread? That's easy to say but probably difficult to implement...and we may run into other unexpected issues

If you search for `moveToThread` in MinimalScene.cc, e.g. [here](https://github.com/ignitionrobotics/ign-gui/blob/6d3690db7cb072664a7382752654763c7cee5a34/src/plugins/minimal_scene/MinimalScene.cc#L521), you can see that we create the opengl objects then move them to the rendering thread. So we'll just need to move them to the same thread that the sensors system in doing rendering and we end up with one thread shared by GUI and sensors.

Reply:
This is all very high level idea. Maybe that should be experimented separately and come as a follow-up PR?

Reply:
> one main rendering loop in a thread shared by sensors and gui.

That would be ideal :sparkles: 

> experimented separately and come as a follow-up PR?

That could work, we're working against `main`, so it should be ok to merge one approach and revert it later for another one.